fdbserver/workloads/WatchLeak.actor.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * WatchLeak.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "flow/actorcompiler.h" // must be last include
+
+namespace {
+
+Key getKey(int i) {
+	return Key(format("watchKey/%d", i));
+}
+
+ACTOR static Future<Void> getWatch(Database cx, int index) {
+	state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(cx);
+	state Future<Void> watchFuture;
+	loop {
+		try {
+			watchFuture = tr->watch(getKey(index));
+			wait(tr->commit());
+			break;
+		} catch (Error& e) {
+			wait(tr->onError(e));
+		}
+	}
+	wait(watchFuture);
+	return Void();
+}
+
+} // namespace
+
+class WatchLeakWorkload : public TestWorkload {
+	int numBatches;
+	int numActiveWatches;
+	double delayBetweenBatches;
+	bool reuseKeys;
+
+	ACTOR static Future<Void> _start(WatchLeakWorkload const* self, Database cx) {
+		state std::vector<Future<Void>> activeWatches;
+		state int batchIndex = 0;
+		for (; batchIndex < self->numBatches; ++batchIndex) {
```

Comment:
What's the behaviour of this workload without the fix, does it OOM or time out?

Reply:
It just keeps increasing the number of watches until eventually the client starts polling for changes. We don't expose the behaviour through the watch API, so the test doesn't fail without the fix, we just have to look at metrics to see if the active watch count is behaving as desired.

Reply:
We don't need to do it here, but we should think about how we can make this test do something active to detect a failure in our cleaning up of watches. For example, maybe we could have some post-test check that the number of active watches on the client matches the number expected by the test, etc.